core-rust/core-api-server/src/core_api/errors.rs

```diff
@@ -69,6 +72,23 @@ impl ErrorDetails for LtsTransactionSubmitErrorDetails {
     }
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct InternalServerErrorResponseForPanic;
+
+impl ResponseForPanic for InternalServerErrorResponseForPanic {
+    type ResponseBody = BoxBody;
+
+    fn response_for_panic(
+        &mut self,
+        _panic_payload: Box<dyn Any + Send + 'static>,
+    ) -> Response<Self::ResponseBody> {
+        // Please note that we deliberately do *not*:
+        // - log the panic payload (since the default panic handler already does this);
+        // - include the panic payload in the response (it may contain sensitive details).
+        server_error::<()>("Unexpected server error").into_response()
```

Comment:
We should ideally return a GUID / Trace ID (like we do in the Gateway) so we/they can correlate the error response with our logs of the panic.

(Perhaps on all error responses?)

Happy for us to pursue this in a later PR

Reply:
of course this sounds like a good idea

> Happy for us to pursue this in a later PR

...yup, especially if that's a global change, for all existing error responses.

Reply:
Alright, so I see 2 options:

**A)** _[easy]_ Let's simply always generate a `LogTraceId` when `trace: None`. Then, magically, on errors:
- The user will get: `{"error_type":"Basic","code":500,"message":"Unexpected server error","trace_id":"abcdefgh12345678"}`
- Our logs will get:
```
# 2023-07-20T15:32:48,851 [INFO/EventLoggerModule/BFT no...64vp9cx0t] (EventLoggerModule.java:166) - bft_nxtrnd ... // regular logs, and then suddenly a panic:
thread 'tokio-runtime-worker' panicked at 'a bug happened', core-api-server/src/core_api/handlers/mempool_list.rs:9:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/39f2657d1101b50f9b71ae460b762d330cc8426b/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/39f2657d1101b50f9b71ae460b762d330cc8426b/library/core/src/panicking.rs:64:14
   2: core_api_server::core_api::handlers::mempool_list::handle_mempool_list::{{closure}}::{{closure}}
             at /Users/jakrawcz/repo1/babylon-node/core-rust/core-api-server/src/core_api/handlers/mempool_list.rs:9:9
   ... // snipped
 127: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /Users/jakrawcz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.2/src/runtime/blocking/pool.rs:471:13
2023-07-20T13:32:49.742979Z  WARN core_api_server::core_api::server: path="/core/mempool/list" error=BasicErrorResponse { code: 500, message: "Unexpected server error", trace_id: Some("abcdefgh12345678") }
# 2023-07-20T15:32:49,854 [INFO/EventLoggerModule/BFT no...64vp9cx0t] (EventLoggerModule.java:249) - lgr_commit ... // next logs
```
_(this means: `abcdefgh12345678` is not mentioned in the stack trace printed by a panic hook, but only in a log that should be right next **BUT OFC sometimes it's gonna be ambiguous**)_

**B)** _[too clever]_ Let's install our own panic hook, which will generate a trace ID, and will print it together with the rest of panic info. THEN it will do `thread_local.latest_panic_trace_id = generated_trace_id`, and later our http-level logger will do `trace: Some(thread_local.latest_panic_trace_id)`.
- The user will get the same as in **A)**
- Our logs will get:
```
# 2023-07-20T15:32:48,851 [INFO/EventLoggerModule/BFT no...64vp9cx0t] (EventLoggerModule.java:166) - bft_nxtrnd ... // regular logs, and then suddenly a panic with trace_id:
thread 'tokio-runtime-worker' panicked at 'a bug happened' with trace ID 'abcdefgh12345678': , core-api-server/src/core_api/handlers/mempool_list.rs:9:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/39f2657d1101b50f9b71ae460b762d330cc8426b/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/39f2657d1101b50f9b71ae460b762d330cc8426b/library/core/src/panicking.rs:64:14
   2: core_api_server::core_api::handlers::mempool_list::handle_mempool_list::{{closure}}::{{closure}}
             at /Users/jakrawcz/repo1/babylon-node/core-rust/core-api-server/src/core_api/handlers/mempool_list.rs:9:9
   ... // snipped
 127: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /Users/jakrawcz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.2/src/runtime/blocking/pool.rs:471:13
2023-07-20T13:32:49.742979Z  WARN core_api_server::core_api::server: path="/core/mempool/list" error=BasicErrorResponse { code: 500, message: "Unexpected server error", trace_id: Some("abcdefgh12345678") }
# 2023-07-20T15:32:49,854 [INFO/EventLoggerModule/BFT no...64vp9cx0t] (EventLoggerModule.java:249) - lgr_commit ... // next logs
```
_(this means: `abcdefgh12345678` is mentioned in the stack trace printed by a panic hook, and it should be **99.9% unambiguous**, except for crazy "panic overwrite" scenarios [like, panic inside a drop during stack unwinding] - at least that's how I understand rust docs on panic hook and unwinding)_